File: solutions/ruby/twelve-days/1/twelve_days.rb
class TwelveDays
  # Definimos los datos base de la canción
  DATA = [
    { day: 'first', gift: 'a Partridge in a Pear Tree.' },
    { day: 'second', gift: 'two Turtle Doves' },
    { day: 'third', gift: 'three French Hens' },
    { day: 'fourth', gift: 'four Calling Birds' },
    { day: 'fifth', gift: 'five Gold Rings' },
    { day: 'sixth', gift: 'six Geese-a-Laying' },
    { day: 'seventh', gift: 'seven Swans-a-Swimming' },
    { day: 'eighth', gift: 'eight Maids-a-Milking' },
    { day: 'ninth', gift: 'nine Ladies Dancing' },
    { day: 'tenth', gift: 'ten Lords-a-Leaping' },
    { day: 'eleventh', gift: 'eleven Pipers Piping' },
    { day: 'twelfth', gift: 'twelve Drummers Drumming' }
  ]

  def self.song
    # Generamos estrofas del 1 al 12 y las unimos con dos saltos de línea
    (1..12).map { |n| verse(n) }.join("\n\n") + "\n"
  end

  private

  def self.verse(n)
    # n-1 porque los arreglos empiezan en 0
    day_info = DATA[n - 1]
    
    # Construimos la introducción
    intro = "On the #{day_info[:day]} day of Christmas my true love gave to me: "
    
    # Construimos la lista de regalos acumulada
    gifts = []
    (n - 1).downto(0) do |i|
      gift_text = DATA[i][:gift]
      
      # Si es el primer regalo (i=0) y no es la primera estrofa (n>1), añadimos 'and '
      if i == 0 && n > 1
        gifts << "and #{gift_text}"
      else
        gifts << gift_text
      end
    end

    intro + gifts.join(", ")
  end
end

File: solutions/typescript/twelve-days/1/twelve-days.ts
const DAYS = [
  'first', 'second', 'third', 'fourth', 'fifth', 'sixth',
  'seventh', 'eighth', 'ninth', 'tenth', 'eleventh', 'twelfth'
];

const GIFTS = [
  'a Partridge in a Pear Tree',
  'two Turtle Doves',
  'three French Hens',
  'four Calling Birds',
  'five Gold Rings',
  'six Geese-a-Laying',
  'seven Swans-a-Swimming',
  'eight Maids-a-Milking',
  'nine Ladies Dancing',
  'ten Lords-a-Leaping',
  'eleven Pipers Piping',
  'twelve Drummers Drumming'
];

/**
 * Recita los versos de "The Twelve Days of Christmas" para un rango específico.
 * * @param start - El día inicial (1-12)
 * @param end - El día final (1-12)
 * @returns Un único string con los versos solicitados perfectamente formateados.
 */
export function recite(start: number, end: number): string {
  const verses: string[] = [];
  
  // Iteramos en el rango solicitado por los argumentos (convertido a índices 0-11)
  for (let i = start - 1; i < end; i++) {
    const day = DAYS[i];
    const giftList: string[] = [];
    
    // Construimos la lista de regalos en orden inverso
    for (let j = i; j >= 0; j--) {
      giftList.push(GIFTS[j]);
    }
    
    let verse: string;
    if (giftList.length === 1) {
      verse = `On the ${day} day of Christmas my true love gave to me: ${giftList[0]}.`;
    } else {
      const lastGift = giftList.pop();
      const firstPart = giftList.join(', ');
      verse = `On the ${day} day of Christmas my true love gave to me: ${firstPart}, and ${lastGift}.`;
    }
    
    // Añadimos el verso con su respectivo salto de línea al final
    verses.push(verse + '\n');
  }
  
  // Unimos todas las cadenas sin añadir caracteres adicionales intermedios
  return verses.join('');
}